### ISSUE TITLE
glTF 2.0: Clamp "scale" of normal texture info between -1.0 and 1.0

### ISSUE THREAD

**Issue by McNopper** [KhronosGroup/glTF#885]
Otherwise, deferred renderer will do have their challenges.

BTW, what is the purpose of this scale (and the other factors, strength for textures)?

**Comment by bghgary** [comment]
They are intended to support variations in textures by reusing the same textures. For example, you can increase or decrease the scale of a normal texture of a material to apply to different rocks in the scene.

**Comment by McNopper** [comment]
Okay, I think I got the point - and I hope I can explain this:

From my understanding, glTF should be the 3D equivalent to PNG or JPG.

When I use photoshop, I can layer PNG images, blend them with values and the result is again - if I export it - e.g. a PNG file. All the blending information etc. is stored inside the PSD file.

In my option, it should be the same with glTF:

I import a glTF file into Unreal Engine, Unity, three,js or any other 3D library. Inside the engine, I can change strength, scale and whatever to an asset, that I do get all the wanted variations (like for rocks).

But, if I export the scene or a single asset to glTF, all these variations and parameters have to be applied.

At the moment, it seems, that the parameters and even the naming from Unity are used and integrated into glTF. Result is, that glTF is too much dependent on Unity. That's good for Unity developers - but not for people using other engines and libraries.

If we keep these strength and scale values like that, it is like we use the PSD file from Photoshop - and not the exported PNG. Of course, information gets lost, but glTf is supposed to be a runtime format, not an interchange format.

I hipe you got my point.

@pjcozzi Can we please rethink/discuss this? If people urgently need this, please make an extension out of it.

**Comment by bghgary** [comment]
@McNopper Can you explain the technical challenges for implementing this? That will help with the discussion.

**Comment by McNopper** [comment]
When rendering deferred, a normalized vector - in this case the normal - can be packed inside a RGB8 texture. Having a scale < -1.0 or > 1.0, this does not work anymore. One has to use a floating point texture, which can be a performance and memory issue. Also, some GPUs do not allow floating point textures as a render target. 

From a design point of view, I do not like the idea to use the artist parameters proposed by Unity. There are many engines (e.g. Unreal Engine 4) and tools (e.g. Substance Painter), which do have different parameter naming and usage. Whenever a glTF asset is loaded by these tools, they do have to mimic the Unity parameter behaviour.

In my opinion, all factors have to be backed into the images, as soon as the glTF asset is exported. From any tool or engine.

**Comment by bghgary** [comment]
Thanks for the explanation. The downside to baking the textures is that the size of the glTF will increase if there are variations. I'm not opposed to this however, especially considering the issues you mentioned.

Is anyone relying on normal scale or occlusion strength? Any issues if we remove them?

**Comment by McNopper** [comment]
I suggest, that we do make an extension out of it, as it makes sense regarding your argument e.g. exporting a Unity scene as glTF.

To explain further my thoughts behind this:
If a 3D model (like a car, human, monster etc.) is exported from Blender or whatever tool, then any factors should be baked inside the occlusion and normal map. If this exported glTF file is openend into the different 3D engines, it should look almost the same (if just IBL with no post processing is applied). In a game or automotive production, this 3D asset can be instantiated several times, with different parameters or even a more sophisticated shader applied.

If the above scene is exported all these factors make sense. But probably more parameters are needed, to exactly describe the 3D scene. And probably this is not possible at all, as glTF cannot describe this.

So, to be on the safe side, let's put it in as an extension.

Also, it is the same with the factors seen inside the PBR materials. I suggest, default values, if a texture is not present.

For instance, with substance painter and designer, you can create really sophistcated materials using nodes and blends and whatever. If non-layered materials are used, the exported, final textures is just a bunch of base color, metallic, roughness, normal, occlusion and other textures. And these can be uses more or less 1:1 by Unity or Unreal. Other similar tools probably do the same export.

**Comment by bghgary** [comment]
McNopper and I just chatted on the phone briefly. The main confusion with the clamping is that the spec does not currently define the formula for the scale factor of normal maps.

Here is the formula for how scale affects the sampled normal:
normalInTangentSpace = vec3(sampledNormal.xy * normalScale, sampledNormal.x)

Note that normals should be normalized before doing final lighting calculations. In deferred rendering, the G-buffer should store the normalized normal vectors after scaling.

**Comment by pjcozzi** [comment]
Who's the best person to help make the final decision here?  @mlimper?

**Comment by mlimper** [comment]
In our original PBR proposal, there were no normal maps, and conceptually I still wouldn't consider them part of the material, but more like a more efficient way to describe the geometry of the surface.
So, we did not consider to have the factors for normal maps at all.

On the one hand, a factor could be handy for normal maps / bump maps in some cases, to allow that their influence is reduced or strenthened. One use case could be be a normal / bump texture being reused to texture multiple objects, having stronger impact on some of them and less impact on others (controlled via the factor). On the other hand, glTF is not a format for content creation, but for delivery... But I feel I am just repeating the last statement of @McNopper within this thread here.

It basically boils down to us having to decide, for each type of texture / map, between:
* Group A: factors, allowing to scale and re-use a texture with modified values in different contexts (also acting as default if no texture is given)
* Group B: no factors, but just default values, which are ignored as soon as a respective texture is present
* Group C: neither factors nor default values, just textures are allowed

How much sense it makes to have a "factor" also seems to depend on the type of map. For height maps for terrain rendering, it seems pretty common to allow for an "exaggeration factor" (@pjcozzi you're the expert - that's correct, right? ;) ... but there, the value "1.0" inside the texture has no real meaning without a context. That is quite different for other maps, where it is well-defined what "1.0" means. My personal classification would be:

*Group A (factor):*
Height / Displacement Map, AO Map*, Metallic Map*, Roughness Map*

*Group B (default value):*
AO Map*, Metallic Map*, Roughness Map*

*Group C (none, just texture):*
Normal Map, AO Map*

*No strong opinion / all mentioned options will be fine.

**Comment by McNopper** [comment]
Had a good talk with Gary and his team brought up a good example: A tree with its leaves. Regarding the leaves, a variation in regards of color alpha, normal scale and occlusion strength makes sense. Even in one single 3D model. This convinced me and I changed my mind. So let's keep the factors inside.

To summarize in GLSL (excluding the PBR, as we agreed on this before):

Normal map and scale:
`normal = M * normalize( (texture(u_normalMap, v_texCoord.st).rgb * 2.0 - 1.0) * vec3(scale, scale, 1.0) );`
M is object to world matrix or whatever space is used.

Ambient occlusion map and strength:
`color = mix(color, color * texture(u_aoMap, v_texCoord.st).r, strength);`


Regarding the height map, I would like to differentiate between a displacement map:
The height map is for parallax mapping.
The displacement map is for displacing geometry by using tessellation and so on.

Both methods possibly need their extra parameters to make sense. Also, both maps can be used in parallel.

**Comment by mlimper** [comment]
> a variation in regards of color alpha, normal scale and occlusion strength makes sense. Even in one single 3D model

OK, then I would be fine with allowing a normal scale as well. Especially since the implementation is very straightforward anyway.

**Comment by McNopper** [comment]
Great! So if everybody agrees, I will close this issue.

**Comment by McNopper** [comment]
Just give me some +1. whoe where involved in this discussion :-)

**Comment by mlimper** [comment]
+1

**Comment by bghgary** [comment]
+1 We will add the formulas to the spec.

**Comment by McNopper** [comment]
That was fast!

**Comment by McNopper** [comment]
Maybe someone can add me as a minor contributor to glTF now :-)
